assertion/function/analyzer.go

```diff
@@ -241,6 +257,175 @@ func run(pass *analysis.Pass) (result interface{}, _ error) {
+	argLoc := util.PosToLocation(argExpr.Pos(), pass)
+
+	// Create the duplicated full trigger
+	// TODO: we just copy the pointer for producer and consumer because I don't see a problem when
+	//  two full triggers share a producer or consumer. We do deep duplication for the param or
+	//  return related producer/consumer, i.e., FuncParam, FuncReturn, ArgPass, UseAsReturn, and I
+	//  don't see other conditional producer/consumer that can be shared between two call sites.
+	//  If we did see such cases in the future, we would want to add a deep copy function for every
+	//  ProduceTrigger or ConsumeTrigger type and make the deep copy here. In this case, we would
+	//  also want to see if it is OK to share the underlying site of the producer/consumer in
+	//  inference engine because we would not want to see a conflict at this site due to different
+	//  call sites.
+	dupTrigger := annotation.FullTrigger{
+		Producer:   trigger.Producer,
+		Consumer:   trigger.Consumer,
+		Controller: nil,
+	}
+	if isParamProducer {
+		dupTrigger.Producer = annotation.DuplicateParamProducer(trigger.Producer, argLoc)
+	}
+	if isReturnConsumer {
+		retLoc := util.PosToLocation(callExpr.Pos(), pass)
+		dupTrigger.Consumer = annotation.DuplicateReturnConsumer(trigger.Consumer, retLoc)
+		// Set up the site that controls the controlled full trigger to be created
+		c := annotation.NewArgKey(callee, 0, argLoc)
+		dupTrigger.Controller = &c
+	}
+
+	return dupTrigger
+}
+
+// findCallsToContractedFunctions finds all the calls to the contracted functions in the given
+// function, and returns a map from every called contracted function to the call expressions that
+// call it.
+func findCallsToContractedFunctions(
+	funcNode *ast.FuncDecl,
+	pass *analysis.Pass,
+	functionContracts functioncontracts.Map,
+) map[*types.Func][]*ast.CallExpr {
+	calls := map[*types.Func][]*ast.CallExpr{}
+	ast.Inspect(funcNode, func(n ast.Node) bool {
+		callExpr, ok := n.(*ast.CallExpr)
+		if !ok {
+			return true
+		}
+
+		ident := util.FuncIdentFromCallExpr(callExpr)
+		if ident == nil {
+			return true
+		}
+
+		funcObj, ok := pass.TypesInfo.ObjectOf(ident).(*types.Func)
+		if !ok {
+			return true
+		}
+
+		// TODO: for now we find the functions with only a single contract nonnil -> nonnil. If we
+		//  want to support multiple contracts or contracts with multiple/other values not only we
+		//  should update here, but we should also make changes to other parts of duplicating
+		//  triggers.
+		if !hasOnlyNonNilToNonNilContract(functionContracts, funcObj) {
+			return true
+		}
+		calls[funcObj] = append(calls[funcObj], callExpr)
+		return true
+	})
+	return calls
+}
+
+// hasOnlyNonNilToNonNilContract returns whether the given function has only one contract that is
+// nonnil->nonnil.
+func hasOnlyNonNilToNonNilContract(funcContracts functioncontracts.Map, funcObj *types.Func) bool {
+	contracts, ok := funcContracts[funcObj]
+	if !ok || len(contracts) != 1 {
```

Comment:
Do we need to restrict ourselves to there being no other contracts, or can we loop through them and just return `true` if any of the contracts present is `nonnil -> noninil`?

Reply:
I am not clear about what we should do with other contracts co-existing with `nonnil->nonnil`. In the future if we want to support multiple contracts, we should think of this problem more comprehensively and more likely we will not have this `hasOnlyNonNilToNonNilContract` function at all. Thus, it should be good to just have this function at the exact restriction level as we support at the current stage.

Reply:
Fair enough. My intuition is that each `// contract` line will be interpreted independently in `AND` fashion, since that's what happens with multiple contracts in Java, but you are right this needs more thought and I am fine with the conservative code for now. Maybe then as a short flow up PR, let's have NilAway warn whenever any function has multiple contracts? Since otherwise they silently suppress one another.